Fixes The order of In-progress lessons.

sahiljoster32 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: The order of In-progress lessons is changed, so that latest exploration come first.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

https://user-images.githubusercontent.com/12553576/153010816-9b8c68a0-7961-4912-9391-56addce0c895.mp4


https://user-images.githubusercontent.com/12553576/153010828-b62bfa11-26b8-456e-83e8-dcf66640fb15.mp4




<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly. Please also include videos/screenshots of the developer tools browser console, so that we can be sure that there are no console errors.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

Co-authored-by: Vasa Srijan Reddy <redvasa@amazon.com>

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
fixed ordering

sahiljoster32 commented:
@nithusha21 PTAL!

nithusha21 reviewed (changes requested):
Hmm, few questions from me:
- Why is this necessary? 
- This approach assumes the input list is sorted based on time of addition to the list? Is that a valid assumption?
- Why do the temp variables you create need to be this.variable? You won't need those variables on the object.

sahiljoster32 committed:
Merge remote-tracking branch 'upstream/develop' into develop

sahiljoster32 committed:
Merge branch 'oppia:develop' into community

sahiljoster32 committed:
Merge remote-tracking branch 'origin/community' into community

sahiljoster32 committed:
fixed temp variable

sahiljoster32 commented:
> Hmm, few questions from me:
> 
> * Why is this necessary?
> * This approach assumes the input list is sorted based on time of addition to the list? Is that a valid assumption?
> * Why do the temp variables you create need to be this.variable? You won't need those variables on the object.

* To show the latest completed and incomplete explorations first.
* Whenever we complete an exploration, it appends the exploration to the list `totalCompletedLessonsList`. In this manner exploration which is completed recently appears to be last in list. So, to show the recent exploration we just have to reverse the list. Similarly, This reverse method is also applied for `totalIncompleteLessonsList`
* temp variables removed from object.

sahiljoster32 commented:
@nithusha21 PTAL!

nithusha21 commented:
> Whenever we complete an exploration, it appends the exploration to the list totalCompletedLessonsList. In this manner exploration which is completed recently appears to be last in list. So, to show the recent exploration we just have to reverse the list. Similarly, This reverse method is also applied for totalIncompleteLessonsList

Can you confirm what happens when you reload the page? Then the list is ordered by the backend right? What ordering is used in that case?

sahiljoster32 commented:
> > Whenever we complete an exploration, it appends the exploration to the list totalCompletedLessonsList. In this manner exploration which is completed recently appears to be last in list. So, to show the recent exploration we just have to reverse the list. Similarly, This reverse method is also applied for totalIncompleteLessonsList
> 
> Can you confirm what happens when you reload the page? Then the list is ordered by the backend right? What ordering is used in that case?

https://user-images.githubusercontent.com/12553576/153012229-2844aaf9-6ed4-4a0c-b73a-fbf5437b4da5.mp4

https://user-images.githubusercontent.com/12553576/153012261-0ad2c094-1ea1-435a-ad30-1630b7c27620.mp4

Order of the list is not changing when we reload or open a new tab, this behaviour is observed in both develop and oppia.org. (above attachments are proof for the same)
Also, backend is not using any sorting for storing the explorations, But explorations are pushed one after another in array. So, that's why recently opened exploration appears to be last in array. 
Also thier is no key in exploration object which reflect latest activity time in exploration object. that's why i just reversed the array!

sahiljoster32 commented:
@nithusha21 PTAL!

nithusha21 commented:
Interesting, one last question is - Is this a change in response to a bug report filed? Or is this something you thought of adding and did it?

sahiljoster32 commented:
> Unassigning @sahiljoster32 since a re-review was requested. @sahiljoster32, please make sure you have addressed all review comments. Thanks!

![audit](https://user-images.githubusercontent.com/12553576/153018249-61f04495-2884-4abe-a181-53b010a81711.jpg)

Yeah this issue is mentioned in Learner Dashboard audit(Product Review II) and i also want this to be changed. Because when a user try to reach latest incomplete(or completed) exploration, user has to open show more button and navigate to the end!!

sahiljoster32 commented:
@nithusha21 PTAL!
